x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/users_and_roles_row_actions.tsx

```diff
@@ -5,20 +5,65 @@
  * 2.0.
  */
 
-import React from 'react';
+import React, { useState } from 'react';
 
-import { EuiButtonIcon } from '@elastic/eui';
+import { EuiButtonIcon, EuiConfirmModal } from '@elastic/eui';
 
-import { MANAGE_BUTTON_LABEL, DELETE_BUTTON_LABEL } from '../constants';
+import { CANCEL_BUTTON_LABEL, MANAGE_BUTTON_LABEL, DELETE_BUTTON_LABEL } from '../constants';
+
+import {
+  REMOVE_ROLE_MAPPING_TITLE,
+  REMOVE_ROLE_MAPPING_BUTTON,
+  REMOVE_USER_BUTTON,
+  ROLE_MODAL_TEXT,
+  USER_MODAL_TITLE,
+  USER_MODAL_TEXT,
+} from './constants';
 
 interface Props {
+  username?: string;
   onManageClick(): void;
   onDeleteClick(): void;
 }
 
-export const UsersAndRolesRowActions: React.FC<Props> = ({ onManageClick, onDeleteClick }) => (
-  <>
-    <EuiButtonIcon onClick={onManageClick} iconType="pencil" aria-label={MANAGE_BUTTON_LABEL} />{' '}
-    <EuiButtonIcon onClick={onDeleteClick} iconType="trash" aria-label={DELETE_BUTTON_LABEL} />
-  </>
-);
+export const UsersAndRolesRowActions: React.FC<Props> = ({
+  onManageClick,
+  onDeleteClick,
+  username,
+}) => {
+  const [deleteModalVisible, setVisible] = useState(false);
+  const showDeleteModal = () => setVisible(true);
+  const closeDeleteModal = () => setVisible(false);
+  const title = username ? USER_MODAL_TITLE(username) : REMOVE_ROLE_MAPPING_TITLE;
+  const text = username ? USER_MODAL_TEXT : ROLE_MODAL_TEXT;
+  const confirmButton = username ? REMOVE_USER_BUTTON : REMOVE_ROLE_MAPPING_BUTTON;
+
+  const deleteModal = (
+    <EuiConfirmModal
+      title={title}
+      onCancel={closeDeleteModal}
+      onConfirm={() => {
+        onDeleteClick();
+        closeDeleteModal();
+      }}
+      cancelButtonText={CANCEL_BUTTON_LABEL}
+      confirmButtonText={confirmButton}
+      buttonColor="danger"
+      defaultFocusedButton="confirm"
+    >
+      <p>{text}</p>
+    </EuiConfirmModal>
+  );
+
+  return (
+    <>
+      {deleteModalVisible && deleteModal}
+      <EuiButtonIcon
+        onClick={onManageClick}
+        iconType="pencil"
+        aria-label={MANAGE_BUTTON_LABEL}
+      />{' '}
+      <EuiButtonIcon onClick={showDeleteModal} iconType="trash" aria-label={DELETE_BUTTON_LABEL} />
```

Comment:
[optional] this should likely be `title="..."` and not just an `aria-label` - EUI auto-sets an aria-label based on the title prop anyway, and title gives an extra UX hint to mouse users who hover over the icon.

Also want to add it looks like the `{' '}` isn't really separating the two icons, so you'd probably want an EuiFlexGroup instead.

But the above are moot points anyway if we eventually refactor this to be an EUI actions column, so feel free to skip this comment

Reply:
Added to Google doc for tech debt after 7.14 FF